File: node/src/query_builder.js
/**
 * VedaDB Node.js Driver - Fluent Query Builder
 *
 * Chainable API for constructing SQL queries programmatically.
 * Supports SELECT, INSERT, UPDATE, DELETE with conditions, joins,
 * ordering, pagination, and aggregations.
 */

'use strict';

/**
 * Escape a value for safe SQL inclusion.
 * @param {*} v
 * @returns {string}
 */
function esc(v) {
  if (v === null || v === undefined) return 'NULL';
  if (typeof v === 'boolean') return v ? 'TRUE' : 'FALSE';
  if (typeof v === 'number') return String(v);
  if (typeof v === 'string') return `'${v.replace(/'/g, "''")}'`;
  if (v instanceof Date) return `'${v.toISOString()}'`;
  if (typeof v === 'object') return `'${JSON.stringify(v).replace(/'/g, "''")}'`;
  return String(v);
}

/**
 * Build a WHERE condition string.
 * @param {string} column
 * @param {string} operator
 * @param {*} value
 * @returns {string}
 */
function buildCondition(column, operator, value) {
  switch (operator.toUpperCase()) {
    case 'IS NULL':
    case 'ISNULL':
      return `${column} IS NULL`;
    case 'IS NOT NULL':
    case 'NOTNULL':
      return `${column} IS NOT NULL`;
    case 'IN':
      if (!Array.isArray(value) || value.length === 0) return '1=0';
      return `${column} IN (${value.map(esc).join(', ')})`;
    case 'NOT IN':
      if (!Array.isArray(value) || value.length === 0) return '1=1';
      return `${column} NOT IN (${value.map(esc).join(', ')})`;
    case 'BETWEEN':
      if (!Array.isArray(value) || value.length !== 2) return '1=0';
      return `${column} BETWEEN ${esc(value[0])} AND ${esc(value[1])}`;
    case 'LIKE':
      return `${column} LIKE ${esc(value)}`;
    default:
      return `${column} ${operator} ${esc(value)}`;
  }
}

/**
 * Fluent query builder for VedaDB.
 */
class QueryBuilder {
  /**
   * @param {Object} client - VedaDB client instance
   * @param {string} table - Table name
   */
  constructor(client, table) {
    this._client = client;
    this._table = table;
    this._columns = ['*'];
    this._where = [];
    this._orderBy = [];
    this._groupBy = [];
    this._having = [];
    this._limit = null;
    this._offset = null;
    this._joins = [];
    this._distinct = false;
    this._forUpdate = false;
    this._aggregate = null;
  }

  // -- Column Selection -----------------------------------------------------

  /**
   * Specify columns to select.
   * @param {...string} columns
   * @returns {QueryBuilder}
   */
  select(...columns) {
    if (columns.length === 1 && Array.isArray(columns[0])) {
      this._columns = columns[0];
    } else {
      this._columns = columns;
    }
    return this;
  }

  /**
   * Select distinct rows.
   * @returns {QueryBuilder}
   */
  distinct() {
    this._distinct = true;
    return this;
  }

  /**
   * Add a COUNT(*) aggregation.
   * @param {string} [alias='count']
   * @returns {QueryBuilder}
   */
  count(alias = 'count') {
    this._aggregate = `COUNT(*) AS ${alias}`;
    return this;
  }

  // -- WHERE Conditions -----------------------------------------------------

  /**
   * Add a WHERE condition.
   * @param {string} column
   * @param {string} [operator='=']
   * @param {*} [value]
   * @returns {QueryBuilder}
   */
  where(column, operator, value) {
    if (value === undefined && operator !== undefined) {
      // Two-arg form: where('name', 'Alice') => name = 'Alice'
      value = operator;
      operator = '=';
    }
    this._where.push({ condition: buildCondition(column, operator, value), type: 'AND' });
    return this;
  }

  /**
   * Add an OR WHERE condition.
   * @param {string} column
   * @param {string} [operator='=']
   * @param {*} [value]
   * @returns {QueryBuilder}
   */
  orWhere(column, operator, value) {
    if (value === undefined && operator !== undefined) {
      value = operator;
      operator = '=';
    }
    this._where.push({ condition: buildCondition(column, operator, value), type: 'OR' });
    return this;
  }

  /**
   * Add a WHERE IN condition.
   * @param {string} column
   * @param {Array} values
   * @returns {QueryBuilder}
   */
  whereIn(column, values) {
    this._where.push({ condition: buildCondition(column, 'IN', values), type: 'AND' });
    return this;
  }

  /**
   * Add a WHERE BETWEEN condition.
   * @param {string} column
   * @param {*} start
   * @param {*} end
   * @returns {QueryBuilder}
   */
  whereBetween(column, start, end) {
    this._where.push({ condition: buildCondition(column, 'BETWEEN', [start, end]), type: 'AND' });
    return this;
  }

  /**
   * Add a raw WHERE condition.
   * @param {string} raw
   * @returns {QueryBuilder}
   */
  whereRaw(raw) {
    this._where.push({ condition: raw, type: 'AND' });
    return this;
  }

  // -- Joins ----------------------------------------------------------------

  /**
   * Add an INNER JOIN.
   * @param {string} table
   * @param {string} leftColumn
   * @param {string} rightColumn
   * @returns {QueryBuilder}
   */
  join(table, leftColumn, rightColumn) {
    this._joins.push(`INNER JOIN ${table} ON ${leftColumn} = ${rightColumn}`);
    return this;
  }

  /**
   * Add a LEFT JOIN.
   * @param {string} table
   * @param {string} leftColumn
   * @param {string} rightColumn
   * * @returns {QueryBuilder}
   */
  leftJoin(table, leftColumn, rightColumn) {
    this._joins.push(`LEFT JOIN ${table} ON ${leftColumn} = ${rightColumn}`);
    return this;
  }

  // -- Ordering / Grouping / Pagination -------------------------------------

  /**
   * Add ORDER BY.
   * @param {string} column
   * @param {string} [direction='ASC']
   * @returns {QueryBuilder}
   */
  orderBy(column, direction = 'ASC') {
    this._orderBy.push(`${column} ${direction.toUpperCase()}`);
    return this;
  }

  /**
   * ORDER BY ... DESC shortcut.
   * @param {string} column
   * @returns {QueryBuilder}
   */
  orderByDesc(column) {
    return this.orderBy(column, 'DESC');
  }

  /**
   * Add GROUP BY.
   * @param {...string} columns
   * @returns {QueryBuilder}
   */
  groupBy(...columns) {
    this._groupBy.push(...columns);
    return this;
  }

  /**
   * Add HAVING condition.
   * @param {string} raw
   * @returns {QueryBuilder}
   */
  having(raw) {
    this._having.push(raw);
    return this;
  }

  /**
   * Set LIMIT.
   * @param {number} n
   * @returns {QueryBuilder}
   */
  limit(n) {
    this._limit = n;
    return this;
  }

  /**
   * Set OFFSET.
   * @param {number} n
   * @returns {QueryBuilder}
   */
  offset(n) {
    this._offset = n;
    return this;
  }

  // -- Query Building -------------------------------------------------------

  /**
   * Build the SELECT SQL string.
   * @returns {string}
   */
  toSQL() {
    let sql = 'SELECT ';
    if (this._distinct) sql += 'DISTINCT ';
    if (this._aggregate) {
      sql += this._aggregate;
    } else {
      sql += this._columns.join(', ');
    }
    sql += ` FROM ${this._table}`;

    for (const join of this._joins) {
      sql += ' ' + join;
    }

    if (this._where.length > 0) {
      sql += ' WHERE ';
      sql += this._where.map((w, i) => (i > 0 ? `${w.type} ` : '') + w.condition).join(' ');
    }

    if (this._groupBy.length > 0) {
      sql += ' GROUP BY ' + this._groupBy.join(', ');
    }

    if (this._having.length > 0) {
      sql += ' HAVING ' + this._having.join(' AND ');
    }

    if (this._orderBy.length > 0) {
      sql += ' ORDER BY ' + this._orderBy.join(', ');
    }

    if (this._limit != null) {
      sql += ` LIMIT ${this._limit}`;
    }

    if (this._offset != null) {
      sql += ` OFFSET ${this._offset}`;
    }

    sql += ';';
    return sql;
  }

  // -- Execution ------------------------------------------------------------

  /**
   * Execute the built SELECT query.
   * @returns {Promise<Result>}
   */
  async execute() {
    return this._client.query(this.toSQL());
  }

  /**
   * Execute and return all rows as objects.
   * @returns {Promise<Object[]>}
   */
  async all() {
    const result = await this.execute();
    return result.toObjects();
  }

  /**
   * Execute and return the first row as an object.
   * @returns {Promise<Object|null>}
   */
  async first() {
    const result = await this.limit(1).execute();
    return result.first();
  }

  /**
   * Execute and return a single column's values.
   * @param {string} column
   * @returns {Promise<any[]>}
   */
  async pluck(column) {
    const result = await this.execute();
    return result.pluck(column);
  }

  // -- INSERT / UPDATE / DELETE ---------------------------------------------

  /**
   * Build and execute an INSERT.
   * @param {Object} data - { column: value }
   * @returns {Promise<Result>}
   */
  async insert(data) {
    const cols = Object.keys(data);
    const vals = cols.map(c => esc(data[c]));
    const sql = `INSERT INTO ${this._table} (${cols.join(', ')}) VALUES (${vals.join(', ')});`;
    return this._client.query(sql);
  }

  /**
   * Build and execute a batch INSERT.
   * @param {Object[]} rows
   * @returns {Promise<Result>}
   */
  async insertMany(rows) {
    if (!rows || rows.length === 0) return { rowCount: 0 };
    const cols = Object.keys(rows[0]);
    const values = rows.map(row =>
      '(' + cols.map(c => esc(row[c])).join(', ') + ')'
    ).join(', ');
    const sql = `INSERT INTO ${this._table} (${cols.join(', ')}) VALUES ${values};`;
    return this._client.query(sql);
  }

  /**
   * Build and execute an UPDATE.
   * @param {Object} data - { column: newValue }
   * @returns {Promise<Result>}
   */
  async update(data) {
    const setClause = Object.entries(data)
      .map(([k, v]) => `${k} = ${esc(v)}`)
      .join(', ');
    let sql = `UPDATE ${this._table} SET ${setClause}`;
    if (this._where.length > 0) {
      sql += ' WHERE ' + this._where.map((w, i) => (i > 0 ? `${w.type} ` : '') + w.condition).join(' ');
    }
    sql += ';';
    return this._client.query(sql);
  }

  /**
   * Build and execute a DELETE.
   * @returns {Promise<Result>}
   */
  async delete() {
    let sql = `DELETE FROM ${this._table}`;
    if (this._where.length > 0) {
      sql += ' WHERE ' + this._where.map((w, i) => (i > 0 ? `${w.type} ` : '') + w.condition).join(' ');
    }
    sql += ';';
    return this._client.query(sql);
  }
}

module.exports = { QueryBuilder, esc };

File: rust/src/query_builder.rs
use crate::error::VedaError;
use crate::result::Value;

/// Fluent SQL query builder for VedaDB.
#[derive(Debug, Clone)]
pub struct QueryBuilder {
    operation: Operation,
    table: String,
    columns: Vec<String>,
    values: Vec<Vec<Value>>,
    sets: Vec<(String, Value)>,
    conditions: Vec<Condition>,
    order_by: Vec<OrderClause>,
    limit: Option<usize>,
    offset: Option<usize>,
    joins: Vec<JoinClause>,
    group_by: Vec<String>,
    having: Vec<Condition>,
    returning: Vec<String>,
    distinct: bool,
    for_update: bool,
    raw_sql: Option<String>,
}

#[derive(Debug, Clone)]
enum Operation {
    Select,
    Insert,
    Update,
    Delete,
    Count,
    Exists,
    Raw,
}

#[derive(Debug, Clone)]
struct Condition {
    column: String,
    op: String,
    value: Value,
    conjunction: String, // AND / OR
}

#[derive(Debug, Clone)]
struct OrderClause {
    column: String,
    direction: OrderDirection,
}

#[derive(Debug, Clone)]
enum OrderDirection {
    Asc,
    Desc,
}

#[derive(Debug, Clone)]
struct JoinClause {
    join_type: JoinType,
    table: String,
    on_condition: String,
}

#[derive(Debug, Clone)]
enum JoinType {
    Inner,
    Left,
    Right,
    Full,
}

impl QueryBuilder {
    /// Start a SELECT query.
    pub fn select(table: &str) -> Self {
        QueryBuilder {
            operation: Operation::Select,
            table: table.to_string(),
            columns: Vec::new(),
            values: Vec::new(),
            sets: Vec::new(),
            conditions: Vec::new(),
            order_by: Vec::new(),
            limit: None,
            offset: None,
            joins: Vec::new(),
            group_by: Vec::new(),
            having: Vec::new(),
            returning: Vec::new(),
            distinct: false,
            for_update: false,
            raw_sql: None,
        }
    }

    /// Start an INSERT query.
    pub fn insert(table: &str) -> Self {
        QueryBuilder {
            operation: Operation::Insert,
            table: table.to_string(),
            columns: Vec::new(),
            values: Vec::new(),
            sets: Vec::new(),
            conditions: Vec::new(),
            order_by: Vec::new(),
            limit: None,
            offset: None,
            joins: Vec::new(),
            group_by: Vec::new(),
            having: Vec::new(),
            returning: Vec::new(),
            distinct: false,
            for_update: false,
            raw_sql: None,
        }
    }

    /// Start an UPDATE query.
    pub fn update(table: &str) -> Self {
        QueryBuilder {
            operation: Operation::Update,
            table: table.to_string(),
            columns: Vec::new(),
            values: Vec::new(),
            sets: Vec::new(),
            conditions: Vec::new(),
            order_by: Vec::new(),
            limit: None,
            offset: None,
            joins: Vec::new(),
            group_by: Vec::new(),
            having: Vec::new(),
            returning: Vec::new(),
            distinct: false,
            for_update: false,
            raw_sql: None,
        }
    }

    /// Start a DELETE query.
    pub fn delete(table: &str) -> Self {
        QueryBuilder {
            operation: Operation::Delete,
            table: table.to_string(),
            columns: Vec::new(),
            values: Vec::new(),
            sets: Vec::new(),
            conditions: Vec::new(),
            order_by: Vec::new(),
            limit: None,
            offset: None,
            joins: Vec::new(),
            group_by: Vec::new(),
            having: Vec::new(),
            returning: Vec::new(),
            distinct: false,
            for_update: false,
            raw_sql: None,
        }
    }

    /// Start a COUNT query.
    pub fn count(table: &str) -> Self {
        QueryBuilder {
            operation: Operation::Count,
            table: table.to_string(),
            columns: Vec::new(),
            values: Vec::new(),
            sets: Vec::new(),
            conditions: Vec::new(),
            order_by: Vec::new(),
            limit: None,
            offset: None,
            joins: Vec::new(),
            group_by: Vec::new(),
            having: Vec::new(),
            returning: Vec::new(),
            distinct: false,
            for_update: false,
            raw_sql: None,
        }
    }

    /// Start from raw SQL.
    pub fn raw(sql: &str) -> Self {
        QueryBuilder {
            operation: Operation::Raw,
            table: String::new(),
            columns: Vec::new(),
            values: Vec::new(),
            sets: Vec::new(),
            conditions: Vec::new(),
            order_by: Vec::new(),
            limit: None,
            offset: None,
            joins: Vec::new(),
            group_by: Vec::new(),
            having: Vec::new(),
            returning: Vec::new(),
            distinct: false,
            for_update: false,
            raw_sql: Some(sql.to_string()),
        }
    }

    // --- Column Selection ---

    /// Select specific columns.
    pub fn columns(mut self, cols: &[&str]) -> Self {
        self.columns = cols.iter().map(|s| s.to_string()).collect();
        self
    }

    /// Select a single column.
    pub fn column(mut self, col: &str) -> Self {
        self.columns.push(col.to_string());
        self
    }

    /// Use DISTINCT.
    pub fn distinct(mut self) -> Self {
        self.distinct = true;
        self
    }

    // --- WHERE Clauses ---

    /// Add an AND WHERE condition.
    pub fn where_eq(mut self, column: &str, value: impl Into<Value>) -> Self {
        self.conditions.push(Condition {
            column: column.to_string(),
            op: "=".to_string(),
            value: value.into(),
            conjunction: "AND".to_string(),
        });
        self
    }

    pub fn where_ne(self, column: &str, value: impl Into<Value>) -> Self {
        self.where_op(column, "!=", value)
    }

    pub fn where_gt(self, column: &str, value: impl Into<Value>) -> Self {
        self.where_op(column, ">", value)
    }

    pub fn where_gte(self, column: &str, value: impl Into<Value>) -> Self {
        self.where_op(column, ">=", value)
    }

    pub fn where_lt(self, column: &str, value: impl Into<Value>) -> Self {
        self.where_op(column, "<", value)
    }

    pub fn where_lte(self, column: &str, value: impl Into<Value>) -> Self {
        self.where_op(column, "<=", value)
    }

    pub fn where_like(mut self, column: &str, pattern: &str) -> Self {
        self.conditions.push(Condition {
            column: column.to_string(),
            op: "LIKE".to_string(),
            value: Value::String(pattern.to_string()),
            conjunction: "AND".to_string(),
        });
        self
    }

    pub fn where_in(mut self, column: &str, values: Vec<impl Into<Value>>) -> Self {
        let vals: Vec<Value> = values.into_iter().map(|v| v.into()).collect();
        self.conditions.push(Condition {
            column: column.to_string(),
            op: "IN".to_string(),
            value: Value::Array(vals),
            conjunction: "AND".to_string(),
        });
        self
    }

    pub fn where_null(mut self, column: &str) -> Self {
        self.conditions.push(Condition {
            column: column.to_string(),
            op: "IS NULL".to_string(),
            value: Value::Null,
            conjunction: "AND".to_string(),
        });
        self
    }

    fn where_op(mut self, column: &str, op: &str, value: impl Into<Value>) -> Self {
        self.conditions.push(Condition {
            column: column.to_string(),
            op: op.to_string(),
            value: value.into(),
            conjunction: "AND".to_string(),
        });
        self
    }

    // --- JOIN Clauses ---

    pub fn inner_join(mut self, table: &str, on: &str) -> Self {
        self.joins.push(JoinClause {
            join_type: JoinType::Inner,
            table: table.to_string(),
            on_condition: on.to_string(),
        });
        self
    }

    pub fn left_join(mut self, table: &str, on: &str) -> Self {
        self.joins.push(JoinClause {
            join_type: JoinType::Left,
            table: table.to_string(),
            on_condition: on.to_string(),
        });
        self
    }

    pub fn right_join(mut self, table: &str, on: &str) -> Self {
        self.joins.push(JoinClause {
            join_type: JoinType::Right,
            table: table.to_string(),
            on_condition: on.to_string(),
        });
        self
    }

    // --- ORDER BY ---

    pub fn order_by(mut self, column: &str, desc: bool) -> Self {
        self.order_by.push(OrderClause {
            column: column.to_string(),
            direction: if desc {
                OrderDirection::Desc
            } else {
                OrderDirection::Asc
            },
        });
        self
    }

    pub fn order_by_asc(self, column: &str) -> Self {
        self.order_by(column, false)
    }

    pub fn order_by_desc(self, column: &str) -> Self {
        self.order_by(column, true)
    }

    // --- LIMIT / OFFSET ---

    pub fn limit(mut self, n: usize) -> Self {
        self.limit = Some(n);
        self
    }

    pub fn offset(mut self, n: usize) -> Self {
        self.offset = Some(n);
        self
    }

    // --- INSERT Values ---

    pub fn values(mut self, vals: Vec<Vec<impl Into<Value>>>) -> Self {
        self.values = vals
            .into_iter()
            .map(|row| row.into_iter().map(|v| v.into()).collect())
            .collect();
        self
    }

    pub fn value(mut self, row: Vec<impl Into<Value>>) -> Self {
        self.values.push(row.into_iter().map(|v| v.into()).collect());
        self
    }

    // --- UPDATE Sets ---

    pub fn set(mut self, column: &str, value: impl Into<Value>) -> Self {
        self.sets.push((column.to_string(), value.into()));
        self
    }

    // --- GROUP BY / HAVING ---

    pub fn group_by(mut self, columns: &[&str]) -> Self {
        self.group_by = columns.iter().map(|s| s.to_string()).collect();
        self
    }

    pub fn having(mut self, column: &str, op: &str, value: impl Into<Value>) -> Self {
        self.having.push(Condition {
            column: column.to_string(),
            op: op.to_string(),
            value: value.into(),
            conjunction: "AND".to_string(),
        });
        self
    }

    // --- RETURNING ---

    pub fn returning(mut self, columns: &[&str]) -> Self {
        self.returning = columns.iter().map(|s| s.to_string()).collect();
        self
    }

    /// FOR UPDATE clause.
    pub fn for_update(mut self) -> Self {
        self.for_update = true;
        self
    }

    // --- Build ---

    /// Build the SQL string.
    pub fn build(self) -> Result<String, VedaError> {
        if let Some(raw) = self.raw_sql {
            return Ok(raw);
        }

        let sql = match self.operation {
            Operation::Select => self.build_select(),
            Operation::Insert => self.build_insert(),
            Operation::Update => self.build_update(),
            Operation::Delete => self.build_delete(),
            Operation::Count => self.build_count(),
            Operation::Exists => self.build_exists(),
            Operation::Raw => unreachable!(),
        };

        Ok(sql)
    }

    fn build_select(&self) -> String {
        let mut sql = String::from("SELECT ");

        if self.distinct {
            sql.push_str("DISTINCT ");
        }

        if self.columns.is_empty() {
            sql.push('*');
        } else {
            sql.push_str(&self.columns.join(", "));
        }

        sql.push_str(&format!(" FROM {}", self.table));

        // JOINs
        for join in &self.joins {
            let join_type = match join.join_type {
                JoinType::Inner => "INNER JOIN",
                JoinType::Left => "LEFT JOIN",
                JoinType::Right => "RIGHT JOIN",
                JoinType::Full => "FULL JOIN",
            };
            sql.push_str(&format!(
                " {} {} ON {}",
                join_type, join.table, join.on_condition
            ));
        }

        // WHERE
        self.append_where(&mut sql);

        // GROUP BY
        if !self.group_by.is_empty() {
            sql.push_str(&format!(" GROUP BY {}", self.group_by.join(", ")));
        }

        // HAVING
        if !self.having.is_empty() {
            let clauses: Vec<String> = self
                .having
                .iter()
                .map(|c| format!("{} {} {}", c.column, c.op, c.value))
                .collect();
            sql.push_str(&format!(" HAVING {}", clauses.join(" AND ")));
        }

        // ORDER BY
        if !self.order_by.is_empty() {
            let orders: Vec<String> = self
                .order_by
                .iter()
                .map(|o| {
                    format!(
                        "{} {}",
                        o.column,
                        match o.direction {
                            OrderDirection::Asc => "ASC",
                            OrderDirection::Desc => "DESC",
                        }
                    )
                })
                .collect();
            sql.push_str(&format!(" ORDER BY {}", orders.join(", ")));
        }

        // LIMIT
        if let Some(limit) = self.limit {
            sql.push_str(&format!(" LIMIT {}", limit));
        }

        // OFFSET
        if let Some(offset) = self.offset {
            sql.push_str(&format!(" OFFSET {}", offset));
        }

        // FOR UPDATE
        if self.for_update {
            sql.push_str(" FOR UPDATE");
        }

        sql.push(';');
        sql
    }

    fn build_insert(&self) -> String {
        let mut sql = format!("INSERT INTO {}", self.table);

        if !self.columns.is_empty() {
            sql.push_str(&format!(" ({})", self.columns.join(", ")));
        }

        if !self.values.is_empty() {
            let value_sets: Vec<String> = self
                .values
                .iter()
                .map(|row| {
                    let vals: Vec<String> = row.iter().map(|v| v.to_string()).collect();
                    format!("({})", vals.join(", "))
                })
                .collect();
            sql.push_str(&format!(" VALUES {}", value_sets.join(", ")));
        }

        if !self.returning.is_empty() {
            sql.push_str(&format!(" RETURNING {}", self.returning.join(", ")));
        }

        sql.push(';');
        sql
    }

    fn build_update(&self) -> String {
        let mut sql = format!("UPDATE {} SET ", self.table);

        let sets: Vec<String> = self
            .sets
            .iter()
            .map(|(col, val)| format!("{} = {}", col, val))
            .collect();
        sql.push_str(&sets.join(", "));

        self.append_where(&mut sql);

        if !self.returning.is_empty() {
            sql.push_str(&format!(" RETURNING {}", self.returning.join(", ")));
        }

        sql.push(';');
        sql
    }

    fn build_delete(&self) -> String {
        let mut sql = format!("DELETE FROM {}", self.table);
        self.append_where(&mut sql);

        if !self.returning.is_empty() {
            sql.push_str(&format!(" RETURNING {}", self.returning.join(", ")));
        }

        sql.push(';');
        sql
    }

    fn build_count(&self) -> String {
        let mut sql = format!("SELECT COUNT(*) FROM {}", self.table);
        self.append_where(&mut sql);
        sql.push(';');
        sql
    }

    fn build_exists(&self) -> String {
        let mut sql = format!(
            "SELECT EXISTS(SELECT 1 FROM {})",
            self.table
        );
        self.append_where(&mut sql);
        sql.push(';');
        sql
    }

    fn append_where(&self, sql: &mut String) {
        if !self.conditions.is_empty() {
            let clauses: Vec<String> = self
                .conditions
                .iter()
                .map(|c| {
                    if c.op == "IS NULL" {
                        format!("{} IS NULL", c.column)
                    } else if c.op == "IN" {
                        if let Value::Array(vals) = &c.value {
                            let items: Vec<String> = vals.iter().map(|v| v.to_string()).collect();
                            format!("{} IN ({})", c.column, items.join(", "))
                        } else {
                            format!("{} IN ({})", c.column, c.value)
                        }
                    } else {
                        format!("{} {} {}", c.column, c.op, c.value)
                    }
                })
                .collect();
            sql.push_str(&format!(" WHERE {}", clauses.join(" AND ")));
        }
    }

    /// Build the SQL and the parameters separately for prepared statements.
    pub fn build_prepared(self) -> Result<(String, Vec<Value>), VedaError> {
        // For prepared statements, we'd replace values with ? placeholders
        // This is a simplified version - full implementation would track param positions
        let sql = self.build()?;
        let params: Vec<Value> = self
            .conditions
            .into_iter()
            .map(|c| c.value)
            .collect();
        Ok((sql, params))
    }
}

impl Default for QueryBuilder {
    fn default() -> Self {
        QueryBuilder::select("")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_select_basic() {
        let sql = QueryBuilder::select("users")
            .columns(&["id", "name", "email"])
            .build()
            .unwrap();
        assert_eq!(sql, "SELECT id, name, email FROM users;");
    }

    #[test]
    fn test_select_where() {
        let sql = QueryBuilder::select("users")
            .where_eq("age", 30i64)
            .where_gt("id", 100i64)
            .build()
            .unwrap();
        assert_eq!(
            sql,
            "SELECT * FROM users WHERE age = 30 AND id > 100;"
        );
    }

    #[test]
    fn test_select_join() {
        let sql = QueryBuilder::select("users")
            .columns(&["users.id", "users.name", "orders.total"])
            .inner_join("orders", "users.id = orders.user_id")
            .build()
            .unwrap();
        assert_eq!(
            sql,
            "SELECT users.id, users.name, orders.total FROM users INNER JOIN orders ON users.id = orders.user_id;"
        );
    }

    #[test]
    fn test_insert() {
        let sql = QueryBuilder::insert("users")
            .columns(&["name", "age", "email"])
            .value(vec!["Alice", 30i64, "alice@example.com"])
            .build()
            .unwrap();
        assert_eq!(
            sql,
            "INSERT INTO users (name, age, email) VALUES ('Alice', 30, 'alice@example.com');"
        );
    }

    #[test]
    fn test_update() {
        let sql = QueryBuilder::update("users")
            .set("name", "Bob")
            .set("age", 31i64)
            .where_eq("id", 1i64)
            .build()
            .unwrap();
        assert_eq!(
            sql,
            "UPDATE users SET name = 'Bob', age = 31 WHERE id = 1;"
        );
    }

    #[test]
    fn test_delete() {
        let sql = QueryBuilder::delete("users")
            .where_eq("status", "inactive")
            .build()
            .unwrap();
        assert_eq!(sql, "DELETE FROM users WHERE status = 'inactive';");
    }

    #[test]
    fn test_order_limit() {
        let sql = QueryBuilder::select("users")
            .order_by_desc("created_at")
            .limit(10)
            .offset(20)
            .build()
            .unwrap();
        assert_eq!(
            sql,
            "SELECT * FROM users ORDER BY created_at DESC LIMIT 10 OFFSET 20;"
        );
    }

    #[test]
    fn test_select_distinct() {
        let sql = QueryBuilder::select("users")
            .distinct()
            .column("country")
            .build()
            .unwrap();
        assert_eq!(sql, "SELECT DISTINCT country FROM users;");
    }

    #[test]
    fn test_where_in() {
        let sql = QueryBuilder::select("users")
            .where_in("status", vec!["active", "pending"])
            .build()
            .unwrap();
        assert_eq!(
            sql,
            "SELECT * FROM users WHERE status IN ('active', 'pending');"
        );
    }

    #[test]
    fn test_count() {
        let sql = QueryBuilder::count("users")
            .where_eq("active", true)
            .build()
            .unwrap();
        assert_eq!(sql, "SELECT COUNT(*) FROM users WHERE active = true;");
    }

    #[test]
    fn test_raw_sql() {
        let sql = QueryBuilder::raw("SELECT * FROM users WHERE id = 1").build().unwrap();
        assert_eq!(sql, "SELECT * FROM users WHERE id = 1");
    }
}
